edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet.Docker/CombinedEdgeletConfigProvider.cs

```diff
@@ -62,6 +67,31 @@ static string BindPath(Uri uri)
                 : uri.AbsolutePath;
         }
 
+        static void InjectEdgeAgentLabels(IModule module, CreateContainerParameters createOptions)
+        {
+            // The IModule argument came from the desired properties in a new deployment. For edge agent only,
+            // save its createOptions and env as labels so we can detect changes in future configs.
+            // Note: createOptions and env for each module are generally saved to a store (see the DeploymentConfigInfo
+            // member of Microsoft.Azure.Devices.Edge.Agent.Core.Agent) and we could get them there, but we'd miss the
+            // bootstrap scenario where edge agent is starting for the first time and therefore has no config store.
+            // That's why edge agent is treated differently: its config is persisted in the container's labels before
+            // it ever starts.
+            if (module.Name.Equals(Constants.EdgeAgentModuleName, StringComparison.OrdinalIgnoreCase))
+            {
+                // get createOptions JSON before making any updates to labels
+                string createOptionsJson = JsonConvert.SerializeObject(createOptions);
+
+                var moduleWithDockerConfig = (IModule<DockerConfig>)module; // cast is safe; base impl already checked it
```

Comment:
This seems a bit scary.. I agree this should work in our cases, but is it guaranteed to be true?

Reply:
What's the scary part exactly? The cast? The caller of this helper first calls `base.GetCombinedConfig` which throws if `module` is not `IModule<DockerConfig>`.

Reply:
Its fine for now.. but we should not take a dependency on checks in callers. What if this code path is called from somewhere else?

Reply:
Agreed. Originally this code was written directly into the calling method. Later, when I moved it into a helper routine, this assumption became even more tenuous. Actually, why didn't I just make this helper take `IModule<DockerConfig>` as an argument instead of `IModule`? 🙄